### PR TITLE
ephemeral storage keep alive interval

### PR DESCRIPTION
**Pull request by bridiver**
Resolves https://github.com/brave/brave-browser/issues/14943

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by pes10k — APPROVED** [review]
LGTM

File: chromium_src/chrome/browser/flag_descriptions.cc [hunk 1]
```diff
@@ -62,6 +62,11 @@ const char kBraveSuperReferralDescription[] =
 const char kBraveEphemeralStorageName[] = "Enable Ephemeral Storage";
 const char kBraveEphemeralStorageDescription[] =
     "Use ephemeral storage for third-party frames";
+const char kBraveEphemeralStorageKeepAliveName[] =
+    "Ephermal Storage Keep Alive";
```
Comment (pes10k) [inline chromium_src/chrome/browser/flag_descriptions.cc:66]: s/Ephermal/Ephemeral/

File: chromium_src/chrome/browser/flag_descriptions.cc [hunk 2]
```diff
@@ -62,6 +62,11 @@ const char kBraveSuperReferralDescription[] =
...
+const char kBraveEphemeralStorageKeepAliveDescription[] =
+    "Keep ephermal storage partitions alive for 30 seconds after last tab "
```
Comment (pes10k) [inline chromium_src/chrome/browser/flag_descriptions.cc:68]: same spelling issue, s/ephermal/ephemeral/

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc [hunk 3]
```diff
@@ -12,6 +12,7 @@
 #include "base/hash/md5.h"
 #include "base/no_destructor.h"
 #include "base/optional.h"
+#include "base/threading/thread_task_runner_handle.h"
```
Comment (goodov) [inline browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:15]: It's not necessary to introduce a thread-bound task runner here, a sequence will do fine. It's generally better to use ```SequencedTaskRunner``` and signal other developers that your intentions are not related to an actual thread-bound operation (like some win32 COM-specific stuff).

https://github.com/chromium/chromium/blob/91.0.4450.1/docs/threading_and_tasks.md#posting-to-the-current-thread

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc [hunk 4]
```diff
@@ -32,6 +33,11 @@ namespace {
 constexpr char kSessionStorageSuffix[] = "/ephemeral-session-storage";
 constexpr char kLocalStorageSuffix[] = "/ephemeral-local-storage";
 
+const base::TimeDelta kStorageKeepAliveDelay =
+    net::features::kTimeoutTcpConnectAttemptMax.Get();
+
+base::TimeDelta g_storage_keep_alive_for_testing = base::TimeDelta::Min();
```
Comment (goodov) [inline browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:39]: for tests you can use ```base::test::ScopedFeatureList``` with an [initializer](https://github.com/chromium/chromium/blob/91.0.4450.1/base/test/scoped_feature_list.h#L108-L110) that takes ```FeatureAndParams``` values.

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc [hunk 5]
```diff
@@ -124,10 +130,31 @@ void EphemeralStorageTabHelper::CreateEphemeralStorageAreasForDomainAndURL(
...
+        FROM_HERE,
+        base::BindOnce([](scoped_refptr<content::TLDEphemeralLifetime>
+                              tld_ephemeral_lifetime) {},
+                       tld_ephemeral_lifetime_),
+        g_storage_keep_alive_for_testing.is_min()
+            ? kStorageKeepAliveDelay
+            : g_storage_keep_alive_for_testing);
+  }
+
   tld_ephemeral_lifetime_ = content::TLDEphemeralLifetime::GetOrCreate(
```
Comment (goodov) [inline browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:148]: so each time an instance will be the same, we still want to do ```PostDelayedTask```?
I think it's better to do ```TLDEphemeralLifetime::GetOrCreate```, check if we got a different instance and then run a timer with an old instance if we need to.

File: chromium_src/chrome/browser/flag_descriptions.cc (hunk 2 above)
Comment (goodov) [inline chromium_src/chrome/browser/flag_descriptions.cc:68]: > for 30 seconds

this will be bogus if ```kTimeoutTcpConnectAttemptMax``` value will change at some point.
I think it's safer to use our own feature parameter so we can have some stability/flexibility if chromium change/remove this value.

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc (hunk 5 above)
Comment (bridiver) [reply browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:148]: That gets very complicated if you have multiple tabs open and different types of redirects (like meta refresh). This is much simpler and more reliable because the the ephemeral storage is only destroyed when the scoped refptr count gets to zero so we don't have to worry about debouncing across requests and tabs

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc (hunk 3 above)
Comment (bridiver) [reply browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:15]: I was confused trying to understand what you were saying until I realized I had unintentionally used `ThreadTaskRunnerHandle` instead of `SequencedTaskRunner`  :)

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc (hunk 4 above)
Comment (bridiver) [reply browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:39]: I'm removing `kTimeoutTcpConnectAttemptMax` because we are going to follow-up with a PR to allow multiple options for the timeout

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc (hunk 5 above)
Comment (bridiver) [reply browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:147]: also this only gets called when the storage changes https://github.com/brave/brave-core/pull/8351/files#diff-c50fbda1dae121ba41d87ae616ec4ea89acb79ed5465918fc9ddaa3301e0b6f2R82 but the effect would be the same either way. We want to keep a reference for 30 seconds longer than it would normally live based on the refcount

Comment (goodov) [reply browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:147]: my proposal is just about this:
```
auto new_tld_ephemeral_lifetime = content::TLDEphemeralLifetime::GetOrCreate(...);
if (new_tld_ephemeral_lifetime != tld_ephemeral_lifetime_) {
  PostTask(...);
  tld_ephemeral_lifetime_ = new_tld_ephemeral_lifetime;
}
```
but now I see that we don't recreate the instance if a domain hasn't changed, so the current implementation is fine.

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc [hunk 6]
```diff
@@ -124,10 +129,31 @@ void EphemeralStorageTabHelper::CreateEphemeralStorageAreasForDomainAndURL(
...
+          net::features::kBraveEphemeralStorageKeepAlive)) {
+    // keep the ephemeral storage alive for some time to handle redirects
+    // including meta refresh or other page driven "redirects" that end up back
+    // at the original origin
+    base::SequencedTaskRunnerHandle::Get()->PostDelayedTask(
+        FROM_HERE,
+        base::BindOnce([](scoped_refptr<content::TLDEphemeralLifetime>
+                              tld_ephemeral_lifetime) {},
+                       tld_ephemeral_lifetime_),
+        g_storage_keep_alive_for_testing.is_min()
```
Comment (goodov) [inline browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:142]: also, maybe a nit, but why use ```is_min``` and not ```is_zero```? min is a negative value, not a default one. ```TimeDelta``` has a default constructor that sets an internal value to 0, and it can be checked with ```is_zero``` which is clearer imo.

**Comment by goodov** [comment]
LGTM

**Review by goodov — APPROVED** [review]

File: browser/ephemeral_storage/ephemeral_storage_tab_helper.cc (hunk 6 above)
Comment (bridiver) [reply browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:142]: because 0 might be a value we want to use in testing when the feature flag goes away

Comment (bridiver) [reply browser/ephemeral_storage/ephemeral_storage_tab_helper.cc:142]: `is_min` seemed  more clear as an invalid value vs something that would be set in `SetKeepAliveTimeDelayForTesting`

**Review by mkarolin — APPROVED** [review]
`chromium_src` changes LGTM
